ath79: add support for TP-Link WPA8730 v1

swegener opened the pull request:
The TP-Link WPA8730 v1 is a PLC Wi-Fi extender with the following specs:

- QCA9558 SoC with 128 MiB RAM and 8 MiB Flash
- 2.4 GHz Wi-Fi via SoC, 5 GHz Wi-Fi via QCA988x
- 3 external GBit ports (QCA8337)
- AV1200 PLC interface (QCA7500)

The internal configuration loks like as this device is based on a three
port PLC device, with the SoC and Wi-Fi added afterwards.

The GMII interface of the SoC is connected via a discrete AR8035
Ethernet PHY chip to one of the internal ports of the switch chip. The
GMII interface of the switch chip connects to the PLC chip, which also
manages the switch chip via its MDIO bus. The MDIO bus to the discrete
Ethernet PHY chip is bit-banged by the SoC.

Sadly this means the switch chip cannot be managed by OpenWrt directly.
There is a TP-Link utility that can be used to upload .pib files to the
PLC chip, changing the switch chip configuration, e.g. disabling IPv6
IGMP snooping.

OEM firmware images use a different variant of the v1 header and
flashing OpenWrt currently requires writing the factory image directly
to the flash via SPI at offset 65536. There is a serial console
available, but the bootloader does not support interrupting the boot
process to enter commands or trigger TFTP recovery. Upgrading is then
possible with a regular sysupgrade image.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener committed:
ath79: add support for TP-Link WPA8730 v1

The TP-Link WPA8730 v1 is a PLC Wi-Fi extender with the following specs:

- QCA9558 SoC with 128 MiB RAM and 8 MiB Flash
- 2.4 GHz Wi-Fi via SoC, 5 GHz Wi-Fi via QCA988x
- 3 external GBit ports (QCA8337)
- AV1200 PLC interface (QCA7500)

The internal configuration looks like as this device is based on a three
port PLC device, with the SoC and Wi-Fi added afterwards.

The GMII interface of the SoC is connected via a discrete AR8035
Ethernet PHY chip to one of the internal ports of the switch chip. The
GMII interface of the switch chip connects to the PLC chip, which also
manages the switch chip via its MDIO bus. The MDIO bus to the discrete
Ethernet PHY chip is bit-banged by the SoC.

Sadly this means the switch chip cannot be managed by OpenWrt directly.
There is a TP-Link utility that can be used to upload .pib files to the
PLC chip, changing the switch chip configuration, e.g. disabling IPv6
IGMP snooping.

For installation rename the OpenWrt factory image to upgrade.bin and
upload via OEM web interface.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,164 @@
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
```
This is in the dtsi, it can be dropped here.

swegener replied:
dropped

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,164 @@
+
+&gpio {
+	status = "okay";
+};
```
This can be dropped, it's not disabled anywhere.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -319,6 +319,10 @@ tplink,tl-wpa8630p-v2-eu|\
+tplink,tl-wpa8730-v1)
+	ucidef_set_led_wlan "wlan" "WLAN" "tp-link:green:wlan" "phy1tpt"
+	ucidef_set_led_wlan "wlan5" "WLAN5" "tp-link:green:wlan5" "phy0tpt"
+	;;
```
This can be dropped entirely as it's in the DTS. The LED names are wrong anyway.

swegener replied:
dropped

adschm commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -564,6 +564,17 @@ define Device/tplink_tl-wpa8630p-v2-int
+  DEVICE_MODEL := TL-WPA8730
+  DEVICE_VARIANT := v1
+  IMAGE_SIZE := 8064k
+  TPLINK_BOARD_ID := TLWPA8730V1
```
TPLINK_BOARD_ID is for safeloader, if you use 8mlzma you need TPLINK_HWID

swegener replied:
I've removed it, using the default hwid of 0x0, as that's what in the official image:

Raw dump of the original image on the SPI flash:
```
65536         0x10000         TP-Link firmware header, firmware version: 0.0.0, image version: "", product ID: 0x0, product version: 0, kernel load address: 0x0, kernel entry point: 0x80002000, kernel offset: 0, kernel length: 512, rootfs offset: 985512, rootfs length: 0, bootloader offset: 0, bootloader length: 0
66048         0x10200         LZMA compressed data, properties: 0x5D, dictionary size: 8388608 bytes, uncompressed size: 2847232 bytes
1114112       0x110000        Squashfs filesystem, little endian, version 4.0, compression:lzma, size: 4043374 bytes, 961 inodes, blocksize: 65536 bytes, created: 2017-04-07 04:15:58
```

Matching firmware update image `wpa8730v1_1_0_2_up(170407)-20170407.bin`:
```
56            0x38            TP-Link firmware header, firmware version: 0.0.0, image version: "", product ID: 0x0, product version: -2026897407, kernel load address: 0x0, kernel entry point: 0x0, kernel offset: 0, kernel length: 512, rootfs offset: 1048576, rootfs length: 1049088, bootloader offset: 7142400, bootloader length: 0
568           0x238           TP-Link firmware header, firmware version: 0.0.0, image version: "", product ID: 0x0, product version: 0, kernel load address: 0x0, kernel entry point: 0x80002000, kernel offset: 0, kernel length: 512, rootfs offset: 985512, rootfs length: 0, bootloader offset: 0, bootloader length: 0
1080          0x438           LZMA compressed data, properties: 0x5D, dictionary size: 8388608 bytes, uncompressed size: 2847232 bytes
1049144       0x100238        Squashfs filesystem, little endian, version 4.0, compression:lzma, size: 4043374 bytes, 961 inodes, blocksize: 65536 bytes, created: 2017-04-07 04:15:58
```

adschm commented:
Your commit message lacks flashing instructions.

swegener commented:
While developing I did a recovery SPI flash installation with the generated v1 TP-Link factory firmware file. u-boot accepts the generated file just fine and boots it. I just switched back to the OEM firmware for trying flashing with a factory image from the webinterface.

OEM firmware seem to use some extended variant of the v1 header. The header identifies as v1.1 and has the same layout.
The update files are a bit more different. The firmware with the v1.1 header is nested within another v1.1 header and the update file also contains some additional header.
I'm going to take a deeper look, as I was not yet able to flash the factory image via the webinterface.

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped.

swegener replied:
done

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,156 @@
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca955x.dtsi"
```
This include should be moved in front of the others (just have a look at recently changed files)

swegener replied:
done

adschm commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -564,6 +564,17 @@ define Device/tplink_tl-wpa8630p-v2-int
+  DEVICE_VARIANT := v1
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
+  TPLINK_HWID := 0x87300001
+  IMAGE_SIZE := 8064k
```
I don't think IMAGE_SIZE is used for anything for this device. I'd drop it.

swegener replied:
You're correct, the variable is not used for anything, because v1 doesn't include `check-size` for the images. `IMAGE_SIZE` is defined for `Device/tplink-8mlzma` from which we inherit and it doesn't match. Does it make sense to add `check-size` to the v1 targets?

For now I've removed it, because it's missing from the other images.

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,156 @@
+
+			uboot: partition@0 {
+				label = "u-boot";
+				reg = <0x00000 0x10000>;
```
Please use six digits after "x" as for the later definitions.

swegener replied:
done

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,156 @@
+		compatible = "gpio-leds";
+
+		led_power: power {
+			label = "tp-link:green:power";
```
Drop "tp-link:" prefix for the labels.

swegener replied:
done

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -323,6 +323,9 @@ ath79_setup_interfaces()
 			"0@eth0" "2:lan:3" "3:lan:2" "4:lan:1" "5:lan:4"
 		;;
+	tplink,tl-wpa8730-v1)
+		ucidef_set_interface_lan "eth0" "dhcp"
```
I personally would prefer to remove the "dhcp" here, so the default config is more like what you would expect from OpenWrt.

swegener replied:
In my oppinion it's a device meant to be added to an existing network and should default to DHCP. Upstream also defaults to DHCP.

swegener commented:
I've updated the flashing instruction. The OEM firmware images have a special header, like the recently added TL-WPA8630 v1, which also requires serial or SPI flashing.

mpratt14 commented:
@swegener Is the header situation similar to what's described here?
https://forum.openwrt.org/t/unknown-tp-link-header/77182

Also I am curious, how exactly did you determine that the AR8035 switch is not directly connected to the SOC?

swegener commented:
> @swegener Is the header situation similar to what's described here?
> https://forum.openwrt.org/t/unknown-tp-link-header/77182

No, that looks completely different. For the WP8730 OEM images binwalk is able to identify the TP-Link v1.1 header and it looks like this:

The first 56 bytes contain some new header.
After that there is a v1.1 header, covering the whole image, except the last 1024 bytes. This header splits the firmware image into a kernel part and a rootfs part.
The kernel part itself has another v1.1 header, containing the load address and entry point.

> Also I am curious, how exactly did you determine that the AR8035 switch is not directly connected to the SOC?

For the WPA8730 its pretty obvious as it consists of two PCBs. One contains the power supply, PLC chip and switch chip. The other PCB contains the SoC and WiFi chips and also has the discrete AR8035 Ethernet PHY chip connecting it to the other PCB via pin headers. I then searched through the released GPL firmware code to find the GPIO ports for bit-banging the MDIO bus to the Ethernet PHY.

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,155 @@
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
```
num-cs can probably be dropped.

sixtyfive commented:
Hi all! I have one of these and would love to try putting OpenWRT on it (if it bricks then it bricks). Is it as simple as building OpenWRT with this PR applied and then flashing via serial? I.e. are there UART solder points on one of the PCBs or else, how did you do it, @swegener?

swegener commented:
![tl-wpa8730v1](https://user-images.githubusercontent.com/814471/102931099-04bdc800-449e-11eb-9b85-edea31ac498d.jpg)

As written the device consists of two boards This is an image of the SoC and Wi-Fi board, which is attached to the front half of the device shell.
The serial interface solder points are in the middle of the PCB, labeled TP1 to TP4. I don't remember the order of Vcc, RX, TX, GND from the top of my head.
I used an SPI programmer with test clips to directly flash the image onto the flash chip.

mpratt14 commented:
> I used an SPI programmer with test clips

This is not necessary though correct? @sixtyfive can load the initramfs-kernel.bin using TFTP via the serial console right?

sixtyfive commented:
> > I used an SPI programmer with test clips
> 
> This is not necessary though correct? @sixtyfive can load the initramfs-kernel.bin using TFTP via the serial console right?

Been meaning to buy a test clip for a while anyways. There's more things than just this that it would be useful for :). Thanks for the picture, @swegener!

swegener commented:
> > > I used an SPI programmer with test clips
> > 
> > This is not necessary though correct? @sixtyfive can load the initramfs-kernel.bin using TFTP via the serial console right?
> 
> Been meaning to buy a test clip for a while anyways. There's more things than just this that it would be useful for :). Thanks for the picture, @swegener!

I did another development session yesterday. The serial interface is TP4=Vcc, TP3=GND, TP2=RX, TP1=TX, RX/TX as seen from the device point of view. I was unable to trigger a bootloader command line or trigger TFTP recovery by pushing the reset button. So until we find a way to generate a proper factory.bin that can be uploaded via the OEM webinterface, flashing via SPI seems to be the only option.

sixtyfive commented:
Test clip is on it's way. Will be a few weeks though.

mpratt14 commented:
@swegener IMO, the only thing blocking this from merge is the fact that there is no factory.bin

recent advances in the mtdsplit driver and OKLI kernel loader allow for better handling of unique situations like this via DTS and makefiles. Can you be more specific about the differences between tp-link v1 header and what's required here? probably we can figure it out

example: #4143

swegener commented:
OEM firmware image wpa8730v1_1_0_2_up(170407)-20170407.bin looks like this.

It has a 56 bytes header, see the bottom of this post for information on this.
After that comes a tp-link v1.1 header, that covers the remaining data. Dumped by a patched version of mktplinkfw:

```
File name              : image
File size              : 0x007d0200 /  8192512 bytes
Version 1 Header size  : 0x00000200 /      512 bytes
Header MD5Sum1         : d2 2f 51 8d c5 28 59 79 b2 52 5f 03 a5 2c 3d 90 (ok)
Header MD5Sum2         : 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 (purpose yet unknown, unchecked here)

Vendor name            : TP-LINK Technologies
Firmware version       : ver. 1.1
Hardware ID            : 0x87300001
Hardware Revision      : 0x00000001
Region code            : 0x00000000

Kernel data offset     : 0x00000200 /      512 bytes
Kernel data length     : 0x00100000 /  1048576 bytes
Kernel load address    : 0x00000000
Kernel entry point     : 0x00000000
Rootfs data offset     : 0x00100200 /  1049088 bytes
Rootfs data length     : 0x006cfc00 /  7142400 bytes
Boot loader data offset: 0x00000000 /        0 bytes
Boot loader data length: 0x00000000 /        0 bytes
Total firmware length  : 0x00000000 /        0 bytes
```

OEM uses separate flash partitions for kernel and rootfs, so this makes sense. Kernel load address and entry point are missing, as this header just splits the image into kernel and rootfs and the upgrade process seems to write each part to the corresponding flash partition. 
The rootfs partition is a normal squashfs image. The kernel partition also has a v1.1 header, containing the kernel load address and entry point used for booting:

```
File name              : kernel
File size              : 0x00100000 /  1048576 bytes
Version 1 Header size  : 0x00000200 /      512 bytes
Header MD5Sum1         : 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 (*ERROR*)
          --> expected : b9 bc 34 57 0b 1a d6 0c 18 2d d5 b2 1d 0d fd 7f 
Header MD5Sum2         : 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 (purpose yet unknown, unchecked here)

Vendor name            : TP-LINK Technologies
Firmware version       : ver. 1.1
Hardware ID            : 0x00000000
Hardware Revision      : 0x00000000
Region code            : 0x00000000

Kernel data offset     : 0x00000200 /      512 bytes
Kernel data length     : 0x000f09a8 /   985512 bytes
Kernel load address    : 0x80002000
Kernel entry point     : 0x8020a900
Rootfs data offset     : 0x00000000 /        0 bytes
Rootfs data length     : 0x00000000 /        0 bytes
Boot loader data offset: 0x00000000 /        0 bytes
Boot loader data length: 0x00000000 /        0 bytes
Total firmware length  : 0x00000000 /        0 bytes
```

Both partitions are adjacent, first kernel, then rootfs, and the OpenWrt DTS merges both partitions into the common firmware partition. The generated OpenWrt sysupgrade.bin image contains a regular v1 header, which the bootloader boots fine and tplink-fw and squashfs-split mtd partitions are correctly recognized. 

For generating a factory.bin image I'm pretty sure both v1.1 headers can be generated by the patched version of mktplinkfw, because checksum of the OEM header is computed correctly. What is still open is the initial 56 bytes header, which looks like this:

```
000000 2f 28 f8 ee 4b c8 37 ee 69 00 0d 81 b6 20 63 59  >/(..K.7.i.... cY<
000010 00 7d 02 24 01 4e 45 57 00 00 00 24 00 00 00 00  >.}.$.NEW...$....<
000020 00 00 00 00 00 7d 02 00 10 0f fe 11 89 4c 83 81  >.....}.......L..<
000030 8e bb 45 a7 28 39 e3 1f                          >..E.(9..<
000038
```

Comparing different OEM firmware versions this is what it seems to be:

```
0x00 - 0x0f: completely changes between OEM firmware versions (likely hash or signature)
0x10 - 0x13: size (matches the number of bytes from 0x14 to the end of the file)
0x14 - 0x17: fixed value \x01NEW
0x18 - 0x1b: size (matches the number of bytes from 0x14 to to the end of the 56 byte header)
0x1c - 0x23: fixed value \x00 (likely padding or some unused field)
0x24 - 0x27: size (matches the number of bytes starting after the 56 byte header to the end of the file)
0x28 - 0x37: completely changes between OEM firmware versions (likely hash or signature)
```

So the only unknowns are the 16 bytes blocks at the beginning and at the end. Because of being 128 bit I suspect some kind of md5 hash or signature.

mpratt14 commented:
I am guessing one of the hash is of the header from 0xf to end of header, and the other hash is of everything after the header

sometimes you get a useful error when giving a bad image...maybe ruining one of the hash on an image from OEM, upload and watching the console might give something helpful...

it's a little more complicated than I thought 😆

adschm commented on target/linux/ath79/dts/qca9558_tplink_tl-wpa8730-v1.dts:
```diff
@@ -0,0 +1,154 @@
+		compatible = "gpio-export";
+
+		led_control {
+			gpio-export,name = "tp-link:led:control";
```
I'd replace the value by a simple "led-control".

swegener replied:
replaced

adschm commented:
Please provide flashing instructions that explain the process in a way that somebody who does not already know what to do can do it.

swegener committed:
build: add build command for TP-Link PLC header

Add build command for adding TP-Link PLC header to images.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener commented:
Just a heads up:

After being pointed at Ghidra in another ticket I took a deeper look at the OEM firmware and I'm now able to generate a factory.bin image that can be flashed over the OEM web interface.
The hashes in the initial 56 byte header use the same salting process as the TP-Link v1 header, but with a different salt, and only cover 32 KiB of data. The header and the salt seems to be specific for the PLC devices. In addition there is a 1024 byte country-code footer at the end of the image.

The generation process should also be able to generate a factory.bin for the WPA8630 v1 (commit 7ef7dbaf7057da37afeefa3753bc4d0f0453e51f) and I suspect the current factory.bin does not work.

swegener committed:
firmware-utils: update

Changing PKG_SOURCE_URL is a temporary thing.

Signed-off-by: Sven Wegener <sven.wegener@stealer.net>

swegener commented:
Can someone take a new look?

Generating the firmware images for OEM uploading requires changes to firmware-utils, see https://github.com/swegener/openwrt-firmware-utils/compare/master..tp-link-v11 for the required commits.

tristan-k commented:
Does someone know how to build a OpenWRT image for the device and can explain it?
